privacy-protections/storage-partitioning/helpers/commonTests.js

```diff
@@ -0,0 +1,552 @@
+/* exported testAPIs */
+/* globals FIRST_PARTY_HOSTNAME */
+
+const timeout = 1000; // ms; used for cross-tab communication APIs
+
+function getURL (path, fileType, key) {
+    const url = new URL(`/partitioning/${path}`, window.location.origin);
+    url.searchParams.set('fileType', fileType);
+    url.searchParams.set('key', key);
+    return url;
+}
+
+const sleepMs = (timeMs) => new Promise(
+    (resolve, reject) => setTimeout(resolve, timeMs)
+);
+
+const loadSubresource = async (tagName, url) => {
+    const element = document.createElement(tagName);
+    document.body.appendChild(element);
+    const resultPromise = new Promise((resolve, reject) => {
+        element.addEventListener('load', resolve, { once: true });
+        element.addEventListener('error', reject, { once: true });
+    });
+    element.src = url;
+    try {
+        return await resultPromise;
+    } catch (e) {
+        // some sort of loading error happened
+        return e;
+    }
+};
+
+// Validates storage APIs which return the value of the token stored in
+// the storage container within each test context. E.g.,
+//
+// document.cookie
+//     same-site: [ { "value": "51c69e1b" }, { "value": "51c69e1b" } ]
+//    cross-site: [ { "value": "fd10f7f5d2cf" }, { "value": "fd10f7f5d2cf" } ]
+function validateStorageAPI (sameSites, crossSites, reference, random) {
+    if (
+        (sameSites.every(v => v.error === 'Unsupported')) &&
+        (crossSites.every(v => v.error === 'Unsupported'))
+    ) {
+        return 'unsupported';
+    } else if (
+        (sameSites.every(v => v.error && v.error.endsWith('is deprecated'))) &&
+        (crossSites.every(v => v.error && v.error.endsWith('is deprecated')))
+    ) {
+        return 'unsupported';
+    }
+
+    if (reference.value !== random) {
+        if (reference.value === null && typeof reference.error !== 'undefined') {
+            return 'error';
+        }
+        return 'fail';
+    }
+
+    if (
+        // (!sameSites.length === configurations['same-site'].iterations) ||
+        // (!crossSites.length === configurations['cross-site'].iterations) ||
+        (!sameSites.every(v => v.value === reference.value)) ||
+        (!crossSites.every(v => v.value === crossSites[0].value)) ||
+        (!crossSites.every(v => v.value !== reference.value))
```

Comment:
I understand that having same value cross site is a failure (no partitioning), but why having same value across `sameSites` is a failure? (I might be reading those files in the wrong order)

Reply:
Line 62 will be true if any value returned for the same-site calls isn't equal to the reference value (read in a same-site iframe in the same tab where the values were set). i.e., it will be true if reference = 123, and the values read in the same-site iframes are [123, 456]. Thinking through it, I don't think we want to fail in this condition and should instead return error.

I should add this in a comment, but my intention with the results is:
* API unsupported --> "unsupported"
* An exact set of conditions we expect when partitioned/blocked --> "pass" (ideally this is "blocked" or "partitoned" in the future)
* An exact set of conditions we expect when unpartitioned/unblocked --> "fail"
* Some unknown / unexpected combination of conditions --> "error"

Line 62 returning true should never happen (partitioned or not), and thus should be an error.

Reply:
I tightened up the logic quite a bit here and documented it.